### PR TITLE
Query: Shortern table alias

### LINKED ISSUE #8255: The generated identifiers in multi level queries are too long for SQLServer
Paste into ComplexNavigationsQueryTestBase, run in ComplexNavigationsQuerySqlServerTest:

```C#
[ConditionalFact]
public virtual void Level4_Include()
{
    using (var context = CreateContext())
    {
        var query = context.LevelOne
            .Select(l1 => l1.OneToOne_Required_PK).Where(t => t != null)
            .Select(l2 => l2.OneToOne_Required_PK).Where(t => t != null)
            .Select(l3 => l3.OneToOne_Required_PK).Where(t => t != null)
            .Select(l4 => l4.OneToOne_Required_FK_Inverse.OneToOne_Required_FK_Inverse)
            .Include(l2 => l2.OneToOne_Optional_FK);

        Assert.Equal(10, query.ToList().Count);
    }
}
```

Exception: `The identifier that starts with 'l1.OneToOne_Required_PK.OneToOne_Required_PK.OneToOne_Required_PK.OneToOne_Required_FK_Inverse.OneToOne_Required_FK_Inverse.OneT' is too long. Maximum length is 128.`

### LINKED ISSUE #8284: Re-enable owned entity types query tests
While implementing #8283 , I added `Equals` method on entity types to make it easy to do comparison while verifying results. Doing so resulted in all of the tests in `ComplexNavigationsOwnedQuerySqlServerTest` failing with following error during seeding phase.
```
System.InvalidOperationException : The instance of entity type 'Level1.OneToOne_Required_PK->Level2.OneToOne_Required_PK->Level3.OneToOne_Required_PK->Level4' cannot be tracked because another instance with the key value 'Id:1' is already being tracked. When attaching existing entities, ensure that only one entity instance with a given key value is attached.
	D:\EntityFramework\src\EFCore\ChangeTracking\Internal\IdentityMap.cs(206,0): at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.IdentityMap`1.Add(TKey key, InternalEntityEntry entry)
	D:\EntityFramework\src\EFCore\ChangeTracking\Internal\IdentityMap.cs(191,0): at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.IdentityMap`1.Add(InternalEntityEntry entry)
	D:\EntityFramework\src\EFCore\ChangeTracking\Internal\StateManager.cs(409,0): at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.StartTracking(InternalEntityEntry entry)
	D:\EntityFramework\src\EFCore\ChangeTracking\Internal\InternalEntityEntry.cs(195,0): at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntry.SetEntityState(EntityState oldState, EntityState newState, Boolean acceptChanges)
	D:\EntityFramework\src\EFCore\ChangeTracking\Internal\InternalEntityEntry.cs(82,0): at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntry.SetEntityState(EntityState entityState, Boolean acceptChanges)
	D:\EntityFramework\src\EFCore\ChangeTracking\Internal\EntityGraphAttacher.cs(64,0): at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.EntityGraphAttacher.PaintAction(EntityEntryGraphNode node)
	D:\EntityFramework\src\EFCore\ChangeTracking\Internal\EntityEntryGraphIterator.cs(25,0): at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.EntityEntryGraphIterator.TraverseGraph(EntityEntryGraphNode node, Func`2 handleNode)
	D:\EntityFramework\src\EFCore\ChangeTracking\Internal\EntityEntryGraphIterator.cs(55,0): at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.EntityEntryGraphIterator.TraverseGraph(EntityEntryGraphNode node, Func`2 handleNode)
	D:\EntityFramework\src\EFCore\ChangeTracking\Internal\EntityEntryGraphIterator.cs(55,0): at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.EntityEntryGraphIterator.TraverseGraph(EntityEntryGraphNode node, Func`2 handleNode)
	D:\EntityFramework\src\EFCore\ChangeTracking\Internal\EntityGraphAttacher.cs(33,0): at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.EntityGraphAttacher.AttachGraph(InternalEntityEntry rootEntry, EntityState entityState)
	D:\EntityFramework\src\EFCore\ChangeTracking\Internal\NavigationFixer.cs(185,0): at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.NavigationFixer.NavigationReferenceChanged(InternalEntityEntry entry, INavigation navigation, Object oldValue, Object newValue)
	D:\EntityFramework\src\EFCore\ChangeTracking\Internal\InternalEntityEntryNotifier.cs(135,0): at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntryNotifier.NavigationReferenceChanged(InternalEntityEntry entry, INavigation navigation, Object oldValue, Object newValue)
	D:\EntityFramework\src\EFCore\ChangeTracking\Internal\ChangeDetector.cs(210,0): at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.ChangeDetector.DetectNavigationChange(InternalEntityEntry entry, INavigation navigation)
	D:\EntityFramework\src\EFCore\ChangeTracking\Internal\ChangeDetector.cs(143,0): at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.ChangeDetector.DetectChanges(InternalEntityEntry entry)
	D:\EntityFramework\src\EFCore\ChangeTracking\Internal\ChangeDetector.cs(112,0): at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.ChangeDetector.DetectChanges(IStateManager stateManager)
	D:\EntityFramework\src\EFCore\ChangeTracking\ChangeTracker.cs(162,0): at Microsoft.EntityFrameworkCore.ChangeTracking.ChangeTracker.DetectChanges()
	D:\EntityFramework\src\EFCore\DbContext.cs(320,0): at Microsoft.EntityFrameworkCore.DbContext.TryDetectChanges()
	D:\EntityFramework\src\EFCore\DbContext.cs(298,0): at Microsoft.EntityFrameworkCore.DbContext.SaveChanges(Boolean acceptAllChangesOnSuccess)
	D:\EntityFramework\src\EFCore\DbContext.cs(276,0): at Microsoft.EntityFrameworkCore.DbContext.SaveChanges()
	D:\EntityFramework\src\EFCore.Specification.Tests\TestModels\ComplexNavigationsModel\ComplexNavigationsModelInitializer.cs(22,0): at Microsoft.EntityFrameworkCore.Specification.Tests.TestModels.ComplexNavigationsModel.ComplexNavigationsModelInitializer.Seed(ComplexNavigationsContext context)
	D:\EntityFramework\test\EFCore.SqlServer.FunctionalTests\ComplexNavigationsOwnedQuerySqlServerFixture.cs(46,0): at Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests.ComplexNavigationsOwnedQuerySqlServerFixture.<CreateTestStore>b__7_0()
	D:\EntityFramework\test\EFCore.SqlServer.FunctionalTests\Utilities\SqlServerTestStore.cs(90,0): at Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests.Utilities.SqlServerTestStore.<>c__DisplayClass21_0.<CreateShared>b__0()
	D:\EntityFramework\src\EFCore.Specification.Tests\TestStore.cs(29,0): at Microsoft.EntityFrameworkCore.Specification.Tests.TestStore.CreateShared(String name, Action initializeDatabase)
	D:\EntityFramework\test\EFCore.SqlServer.FunctionalTests\Utilities\SqlServerTestStore.cs(85,0): at Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests.Utilities.SqlServerTestStore.CreateShared(Action initializeDatabase)
	D:\EntityFramework\test\EFCore.SqlServer.FunctionalTests\Utilities\SqlServerTestStore.cs(38,0): at Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests.Utilities.SqlServerTestStore.GetOrCreateShared(String name, Action initializeDatabase, Boolean cleanDatabase)
	D:\EntityFramework\test\EFCore.SqlServer.FunctionalTests\ComplexNavigationsOwnedQuerySqlServerFixture.cs(41,0): at Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests.ComplexNavigationsOwnedQuerySqlServerFixture.CreateTestStore()
	D:\EntityFramework\src\EFCore.Specification.Tests\ComplexNavigationsQueryTestBase.cs(32,0): at Microsoft.EntityFrameworkCore.Specification.Tests.ComplexNavigationsQueryTestBase`2..ctor(TFixture fixture)
	D:\EntityFramework\src\EFCore.Specification.Tests\ComplexNavigationsOwnedQueryTestBase.cs(15,0): at Microsoft.EntityFrameworkCore.Specification.Tests.ComplexNavigationsOwnedQueryTestBase`2..ctor(TFixture fixture)
	D:\EntityFramework\test\EFCore.SqlServer.FunctionalTests\ComplexNavigationsOwnedQuerySqlServerTest.cs(17,0): at Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests.ComplexNavigationsOwnedQuerySqlServerTest..ctor(ComplexNavigationsOwnedQuerySqlServerFixture fixture, ITestOutputHelper testOutputHelper)
```

### PR DESCRIPTION
**Pull request by smitpatel**
Resolves #8255 

Resolves #8284

### CODE REVIEW

File: src/EFCore.Relational/Query/RelationalQueryCompilationContext.cs [hunk 1]
```diff
@@ -142,6 +142,12 @@ public virtual string CreateUniqueTableAlias([NotNull] string currentAlias)
                 return currentAlias;
             }
 
+            while(currentAlias.Length > 125)
```
Comment (anpete) [inline src/EFCore.Relational/Query/RelationalQueryCompilationContext.cs]: Magic number? Make this a provider overrideable property.

**Comment by smitpatel** [comment]
Updated. Removed the changes in result assertion for now.

File: src/EFCore.Relational/Query/RelationalQueryCompilationContext.cs [hunk 2]
```diff
@@ -142,6 +144,12 @@ public virtual string CreateUniqueTableAlias([NotNull] string currentAlias)
                 return currentAlias;
             }
 
+            while(currentAlias.Length > MaxTableAliasLength - 3)
+            {
+                var index = currentAlias.IndexOf(".", StringComparison.OrdinalIgnoreCase);
+                currentAlias = index > 0 ? currentAlias.Substring(index + 1) : currentAlias.Substring(0, 125);
```
Comment (maumar) [inline src/EFCore.Relational/Query/RelationalQueryCompilationContext.cs]: missed 125 here

**Comment by AndriySvyryd** [comment]
The fix for #8284 is to change line 169 in `StateManager.cs` to:
`entries = new Dictionary<object, InternalEntityEntry>(ReferenceEqualityComparer.Instance);`

File: src/EFCore.Relational/Query/RelationalQueryCompilationContext.cs [hunk 3]
```diff
@@ -82,6 +79,11 @@ var relationalQueryModelVisitor
         public virtual bool IsLateralJoinSupported => false;
 
         /// <summary>
+        ///     Max length of the table alias supported by provider.
+        /// </summary>
+        public virtual int MaxTableAliasLength => 128;
```
Comment (maumar) [inline src/EFCore.Relational/Query/RelationalQueryCompilationContext.cs]: nit: I would also change this to maximum length of alias we generate (before the uniquefying happens), so 125 - makes it more flexible for the providers, e.g if they decide 3 chars overhead for uniquefying is not enough

**Comment by AndriySvyryd** [comment]
Don't forget to unskip the tests in `ComplexNavigationsOwnedQuerySqlServerTest`

**Pull request merged**